[AutoDiff] Directly SILGen `@derivative` attributes to diff witnesses.

dan-zheng opened the pull request:
Previously, `@derivative` attribute type-checking created implicit
`@differentiable` attributes on the original declaration. This was a
longstanding hack powering `@derivative` attribute derivative registration.

https://github.com/apple/swift/pull/28608 made these changes:
- Derivative function configurations (from `@differentiable` and `@derivative`
  attributes) are serialized in modules and are loaded from imported modules.
- The differentiation transform uses these derivative function configurations
  for derivative function lookup instead of `@differentiable` attributes.

Now, `@derivative` attributes are directly lowered to differentiability
witnesses during SILGen, and implicit `@differentiable` attribute generation
is removed.

Resolves TF-835.

Unblocks TF-1021: lifting the "same-file derivative registration only"
limitation in `@derivative` attribute type-checking. This should be possible
without much work, but needs testing!

Exposes TF-1037: crash due to no derivative generic signature mangling (TF-680).
Exposes TF-1040: `@differentiable` attribute limitations for class methods.
Exposes TF-1041: untested protocol requirement `@differentiable` attribute
type-checking logic.

dan-zheng committed:
[AutoDiff] Directly SILGen `@derivative` attributes to diff witnesses.

Previously, `@derivative` attribute type-checking created implicit
`@differentiable` attributes on the original declaration. This was a
longstanding hack powering `@derivative` attribute derivative registration.

https://github.com/apple/swift/pull/28608 made these changes:
- Derivative function configurations (from `@differentiable` and `@derivative`
  attributes) are serialized in modules and are loaded from imported modules.
- The differentiation transform uses these derivative function configurations
  for derivative function lookup instead of `@differentiable` attributes.

Now, `@derivative` attributes are directly lowered to differentiability
witnesses during SILGen, and implicit `@differentiable` attribute generation
is removed.

Type-checking changes:
- "Overlapping" `@differentiable` and `@derivative` attributes (for the same
  original declaration and parameter indices) are now disallowed. They
  semantically conflict because the first "requests derivative generation"
  while the second "registers a derivative".
  - "Overlapping" `@differentiable` and `@derivative` attributes are allowed
    for protocol requirements. Requirement `@differentiable` attributes
    mean "add JVP/VJP witness table entries" - not "request derivative
    generation", because there is no function body.
  - Note that relaxing the "overlapping" condition to consider derivative
    generic signatures is possible after derivative generic signature mangling
    for derivative functions: TF-680.

Resolves TF-835.

Unblocks TF-1021: lifting the "same-file derivative registration only"
limitation in `@derivative` attribute type-checking. This should be possible
without much work, but needs testing!

Exposes TF-1040: `@differentiable` attribute limitations for class methods.
Exposes TF-1041: untested protocol requirement `@differentiable` attribute
type-checking logic.

dan-zheng committed:
Update checkout for tensorflow-swift-apis.

Fix "a derivative already exists" error for `Layer.inferring(from:)`.

dan-zheng commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3918,74 +3919,35 @@ void AttributeChecker::visitDerivativeAttr(DerivativeAttr *attr) {
+      originalAFD->isProtocolRequirement();
+  if (diffAttr && !isOriginalAFDProtocolRequirement) {
+    diagnoseAndRemoveAttr(attr,
+                          diag::derivative_attr_original_already_has_derivative,
```
I reused the `derivative_attr_original_already_has_derivative ` diagnostic, but it arguably has room for improvement:
```
/swift-base/tensorflow-swift-apis/Sources/TensorFlow/Layer.swift:94:6: error: a derivative already exists for 'inferring(from:)'
    @derivative(of: inferring(from:))
    ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    
/swift-base/tensorflow-swift-apis/Sources/TensorFlow/Layer.swift:88:6: note: other attribute declared here
    @differentiable
     ^
```
Showing `"a derivative already exists"` for a `@differentiable` attribute may not be clear to users who don't know that `@differentiable` attribute triggers derivative function generation.

I'm open to improvement suggestions, if anyone has ideas. Otherwise, I'll defer 'til later.

dan-zheng replied:
Obsolete because this diagnostic is no longer being emitted. `@differentiable` + `@derivative` attributes are never in conflict: https://github.com/apple/swift/pull/28621#issuecomment-562751760

dan-zheng committed:
Clarify comment in `TBDGenVisitor::addDerivativeAttr`.

rxwei commented:
> "Overlapping" `@differentiable` and `@derivative` attributes (for the same
original declaration and parameter indices) are now disallowed.

This doesn't make sense to me. 

In protocol conformances, it is perfectly legal to declare a conformance and then provide implementations in a different extension in the same module.  Differentiable functions should behave the same way.

rxwei commented:
> They semantically conflict because the first "requests derivative generation" while the second "registers a derivative".

The semantics of `@differentiable` is not requesting derivative generation. It is declaring a function to be differentiable. With a declaration, the compiler then tries to verify it, either by _differentiating_ it, or by checking whether there's a derivative (`@derivative(of:)`) already.

dan-zheng commented:
> The semantics of `@differentiable` is not requesting derivative generation. It is declaring a function to be differentiable. With a declaration, the compiler then tries to verify it, either by _differentiating_ it, or by checking whether there's a derivative (`@derivative(of:)`) already.

That's a good point, I think I agree! This type-checking change seems misguided - the original motivation was to avoid a crash regarding "`@differentiable` + `@derivative` attribute for same original declaration and parameter indices but different derivative generic signature".

Reproducer:
```swift
protocol P: Differentiable {}
extension P {
  @differentiable
  func foo() -> Float { 1 }
}

extension P {
  @derivative(of: foo)
  func vjpFoo() -> (value: Float, pullback: (Float) -> (TangentVector)) {
    fatalError()
  }
}
```

Previously, `@derivative` attribute type-checking [looked for a `@differentiable` attribute with the same parameter indices and filled in the JVP/VJP](https://github.com/apple/swift/blob/ee7644dd102d7125ecc2da35eb117d7b13866d18/lib/Sema/TypeCheckAttr.cpp#L3923).

```
// differentiability witness for P.foo()
sil_differentiability_witness hidden [parameters 0] [results 0] @$s1p1PPAAE3fooSfyF : $@convention(method) <Self where Self : P> (@in_guaranteed Self) -> Float {
  vjp: @AD__$s1p1PPAAE3fooSfyF__vjp_src_0_wrt_0 : $@convention(method) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> (Float, @owned @callee_guaranteed (Float) -> @out τ_0_0.TangentVector)
}
```

This generated a single differentiability witness, no problem.

---

However, with the current logic in this patch: the `@differentiable` attribute has no derivative generic signature, but the `@derivative` attribute has a derivative generic signature (the generic signature of the `@derivative` declaration is used, which is `<Self where Self : P>`).

This causes two differentiability witnesses to be SILGen'd, each with a different derivative generic signature:
```
// differentiability witness for P.foo()
sil_differentiability_witness hidden [parameters 0] [results 0] @$s1p1PPAAE3fooSfyF : $@convention(method) <Self where Self : P> (@in_guaranteed Self) -> Float {
}

// differentiability witness for P.foo()
sil_differentiability_witness hidden [parameters 0] [results 0] <Self where Self : P> @$s1p1PPAAE3fooSfyF : $@convention(method) <Self where Self : P> (@in_guaranteed Self) -> Float {
  vjp: @AD__$s1p1PPAAE3fooSfyF__vjp_src_0_wrt_0 : $@convention(method) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> (Float, @owned @callee_guaranteed (Float) -> @out τ_0_0.TangentVector)
}
```

Since derivative generic signature mangling is not yet implemented for derivative functions (TF-680), there's a name conflict issue for derivative functions with the same parameter indices but different derivative generic signatures.

SILGen generates a VJP thunk for the second differentiability witness above called `s1p1PPAAE3fooSfyF__vjp_src_0_wrt_0`. The differentiation transform attempts to create an empty VJP function for the first differentiability witness above, but encounters a name clash:
```
// AD__$s1p1PPAAE3fooSfyF__vjp_src_0_wrt_0
sil hidden [thunk] [always_inline] [ossa] @AD__$s1p1PPAAE3fooSfyF__vjp_src_0_wrt_0 : $@convention(method) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> (Float, @owned @callee_guaranteed (Float) -> @out τ_0_0.TangentVector) {
// %0                                             // user: %2
bb0(%0 : $*τ_0_0):
  // function_ref P.vjpFoo()
  %1 = function_ref @$s1p1PPAAE6vjpFooSf5value_13TangentVectorQzSfc8pullbacktyF : $@convention(method) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> (Float, @owned @callee_guaranteed (Float) -> @out τ_0_0.TangentVector) // user: %2
  %2 = apply %1<τ_0_0>(%0) : $@convention(method) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> (Float, @owned @callee_guaranteed (Float) -> @out τ_0_0.TangentVector) // user: %3
  return %2 : $(Float, @callee_guaranteed (Float) -> @out τ_0_0.TangentVector) // id: %3
} // end sil function 'AD__$s1p1PPAAE3fooSfyF__vjp_src_0_wrt_0'

Assertion failed: (!entry->getValue() && "function already exists"), function create, file /Users/danielzheng/swift-bart/swift/lib/SIL/SILFunction.cpp, line 74.
Stack dump:
0.	Program arguments: /Users/danielzheng/swift-bart/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/bin/swift -frontend -interpret p.swift -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -color-diagnostics -module-name p
1.	Swift version 5.1.1-dev (Swift f70940798d)
2.	While running pass #31 SILModuleTransform "Differentiation".
3.	While processing // differentiability witness for P.foo()
sil_differentiability_witness hidden [parameters 0] [results 0] @$s1p1PPAAE3fooSfyF : $@convention(method) <Self where Self : P> (@in_guaranteed Self) -> Float {
}
```

---

According to [the semantics of `@differentiable`](https://github.com/apple/swift/pull/28621#issuecomment-562751760), the correct fix seems to be to "find a way to continue generating a single differentiability witness" for this program. I briefly tried various techniques without success. I can try again, if we agree this is the right approach.

---

Derivative generic signature mangling (TF-860) is not a robust fix for this program because, with the current logic, it would actually trigger VJP generation for the first differentiability witness, which seems surprising (`P.foo` may not even be differentiable).

Current crashers like TF-1037 should be fixed by creating multiple differentiability witnesses when `@differentiable` and `@derivative` derivative generic signatures are truly different.

rxwei commented:
If there's a parameter indices mismatch or generic signature mismatch between a `@differentiable` and a `@derivative(of:)`, they should be treated as being completely separate. 

If parameter indices and generic signatures do match, then preventively check whether such a differentiability witness already exists before adding a new one.

marcrasi commented on include/swift/AST/ASTContext.h:
```diff
@@ -290,11 +292,26 @@ class ASTContext final {
+  llvm::DenseMap<
+      std::tuple<Decl *, IndexSubset *, AutoDiffDerivativeFunctionKind>,
+      DerivativeAttr *>
+      DerivativeAttrs;
```
Would it be possible to use the derivative config list that you created in the previous serialization PR, so that we do not have to maintain another list?

Could the derivative config list actually also replace `DifferentiableAttrs`?

dan-zheng replied:
I wondered the same. The derivative function configuration doesn't currently store whether the configuration came from a `@differentiable` attribute or `@derivative` attribute, and that information is important for users of `DifferentiableAttrs ` and `DerivativeAttrs` (checking duplicate attributes of a specific kind, not just configurations).

I haven't thought deeply about this. I'll file an issue tracking this question if it isn't resolved by the time this PR is merged.

marcrasi commented on lib/TBDGen/TBDGen.cpp:
```diff
@@ -232,14 +233,46 @@ void TBDGenVisitor::addDifferentiabilityWitness(
+                              attr->getDerivativeGenericSignature());
+}
+
+void TBDGenVisitor::addDerivativeAttr(AbstractFunctionDecl *derivative,
```
This this be simplified by iterating over the list of derivative configurations from the previous PR, instead of iterating over the attributes?

I'm not totally sure if we're allowed to use that information in TBDGen. I think it's allowed. For example, TBDGen uses `DC->getLocalConformances()` which uses `NominalTypeDecl::ConformanceTable`, which is a similar piece of information.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/28621/commits/0c531bb8f21a556e367a5206f932d8980570291f. TBDGen.cpp is simpler now!

dan-zheng committed:
Address review feedback.

- Revert `@differentiable` + `@derivative` attribute type-checking changes.
  `@differentiable` + `@derivative` attribute with the same original declaration
  and parameter indices are not in conflict.
- Simplify TBDGen for AutoDiff symbols using
  `AbstractFunctionDecl::getDerivativeFunctionConfigurations`.
- Add TF-1037 negative tests.

dan-zheng committed:
Improve test name.

dan-zheng commented:
Extended test suite fails weirdly:
```
/swift-models/FastStyleTransfer/Models/TransformerNet.swift:49:32: error: type of expression is ambiguous without more context
        let convolved1 = input.sequenced(through: conv1, in1, relu)
                         ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/swift-models/FastStyleTransfer/Models/TransformerNet.swift:91:22: error: type of expression is ambiguous without more context
        return input.sequenced(through: reflectionPad, conv2d)
               ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/swift-models/FastStyleTransfer/Models/TransformerNet.swift:125:30: error: type of expression is ambiguous without more context
        return input + input.sequenced(through: conv1, in1, relu, conv2, in2)
                       ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/swift-models/FastStyleTransfer/Models/TransformerNet.swift:174:29: error: type of expression is ambiguous without more context
        return resizedInput.sequenced(through: reflectionPad, conv2d)
               ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This may be because this PR updates the commit hash for `tensorflow/swift-apis`. The `type of expression is ambiguous without more context` errors may be unrelated to this patch - I'll create a PR that just updates `tensorflow/swift-apis` to verify.

dan-zheng committed:
Fix `@differentiable` attribute requirement type-checking in non-primary files.

dan-zheng commented:
`type of expression is ambiguous without more context` errors fixed in https://github.com/apple/swift/pull/28621/commits/571e9d9692a0d4caeefc7b650ff87520725cdff6.

The fix is: protocol witness `@differentiable` attributes in non-primary-files need to be type-checked via `DifferentiableAttr::getParameterIndices` so that the witness derivative configurations are updated. This makes sense, and similar logic exists in `SILFunctionBuilder::addFunctionAttributes`.

---

Ideally, the diagnostics should have read `candidate is missing attribute '@differentiable'`. Poor diagnostics involving associated type inference are tracked by TF-1014. I investigated TF-1014 and left some details in a comment.

dan-zheng commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -533,55 +533,75 @@ swift::matchWitness(
-        auto implicitDiffAttr = false;
-        if (reqDiffAttrSupersetMatch) {
-          auto *witnessAFD = cast<AbstractFunctionDecl>(witness);
+      (void)reqDiffAttr->getParameterIndices();
```
Note: I found that calling `DifferentiableAttr::getParameterIndices` on witness attributes was sufficient to make `test/AutoDiff/differentiable_attr_type_checking_primary_file.swift` pass. For good measure, I called it on requirement attributes too.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng committed:
Implement `@derivative` attribut serialization.

Fix test/Serialization/derivative_attr.swift.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

marcrasi commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3918,75 +3915,18 @@ void AttributeChecker::visitDerivativeAttr(DerivativeAttr *attr) {
-    }
-    originalAFD->getAttrs().add(da);
+  // Valid `@derivative` attributes are uniqued by original function and
+  // parameter indices. Reject duplicate attributes.
```
This comment is confusing because it doesn't mention `kind` which is also in the uniquification key. I suggest just saying "Reject duplicate attributes" because that will never go stale even if we change the uniquification key in the future. The person reading the comment can easily look at the key in the code to see what it's uniqued by.

dan-zheng replied:
Good catch! Done in https://github.com/apple/swift/pull/28621/commits/f9ced25561c3bbb41ca628a9e41f72f13137562c.

dan-zheng committed:
Address review feedback.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
Does this implementation handle duplicate but private-to-separate-files derivatives?

dan-zheng commented:
> Does this implementation handle duplicate but private-to-separate-files derivatives?

I believe the `@derivative` attribute SILGen logic should work correctly for that case: the SILModule for each file should have its own differentiability witness.

The same-file derivative registration limitation hasn't been lifted from Sema yet (TF-1021), but I verified that it can be lifted trivially after this patch:
* https://github.com/dan-zheng/swift/commit/enable-retrodiff (three file test)
* https://github.com/dan-zheng/swift/commit/enable-retrodiff-duplicate (four file test, duplicate attribute)

Lifting the restriction requires more testing and will be the next follow-up after this patch.

dan-zheng commented:
This patch causes a [fastai/fastai_dev](https://github.com/fastai/fastai_dev) breakage:
```
// AD__$s27FastaiNotebook_07_batchnorm22LearningPhaseDependentPAAE7forwardy6OutputQz5InputQzF__vjp_src_0_wrt_0_1
sil [always_inline] [ossa] @AD__$s27FastaiNotebook_07_batchnorm22LearningPhaseDependentPAAE7forwardy6OutputQz5InputQzF__vjp_src_0_wrt_0_1 : $@convention(method) <Ï„_0_0 where Ï„_0_0 : LearningPhaseDependent> (@in_guaranteed Ï„_0_0.Input, @in_guaranteed Ï„_0_0) -> (@out Ï„_0_0.Output, @owned @callee_guaranteed (@in_guaranteed Ï„_0_0.Output.TangentVector) -> (@out Ï„_0_0.Input.TangentVector, @out Ï„_0_0.TangentVector)) {
// %0                                             // user: %4
// %1                                             // user: %4
// %2                                             // user: %4
bb0(%0 : $*Ï„_0_0.Output, %1 : $*Ï„_0_0.Input, %2 : $*Ï„_0_0):
  // function_ref LearningPhaseDependent.gradForward(_:)
  %3 = function_ref @$s27FastaiNotebook_07_batchnorm22LearningPhaseDependentPAAE11gradForwardy6OutputQz5value_13TangentVectorQz_5Input_AHQZtAE_AHQZc8pullbacktAJQzF : $@convention(method) <Ï„_0_0 where Ï„_0_0 : LearningPhaseDependent> (@in_guaranteed Ï„_0_0.Input, @in_guaranteed Ï„_0_0) -> (@out Ï„_0_0.Output, @owned @callee_guaranteed (@in_guaranteed Ï„_0_0.Output.TangentVector) -> (@out Ï„_0_0.TangentVector, @out Ï„_0_0.Input.TangentVector)) // user: %4
  %4 = apply %3<Ï„_0_0>(%0, %1, %2) : $@convention(method) <Ï„_0_0 where Ï„_0_0 : LearningPhaseDependent> (@in_guaranteed Ï„_0_0.Input, @in_guaranteed Ï„_0_0) -> (@out Ï„_0_0.Output, @owned @callee_guaranteed (@in_guaranteed Ï„_0_0.Output.TangentVector) -> (@out Ï„_0_0.TangentVector, @out Ï„_0_0.Input.TangentVector)) // user: %6
  // function_ref AD__$s6Output_13TangentVectorQZABQz5Input_ABQZIegnrr_AcfDIegnrr_27FastaiNotebook_07_batchnorm22LearningPhaseDependentRzlTR_pullback_self_reordering_thunk
  %5 = function_ref @AD__$s6Output_13TangentVectorQZABQz5Input_ABQZIegnrr_AcfDIegnrr_27FastaiNotebook_07_batchnorm22LearningPhaseDependentRzlTR_pullback_self_reordering_thunk : $@convention(thin) <Ï„_0_0 where Ï„_0_0 : LearningPhaseDependent> (@in_guaranteed Ï„_0_0.Output.TangentVector, @guaranteed @callee_guaranteed (@in_guaranteed Ï„_0_0.Output.TangentVector) -> (@out Ï„_0_0.TangentVector, @out Ï„_0_0.Input.TangentVector)) -> (@out Ï„_0_0.Input.TangentVector, @out Ï„_0_0.TangentVector) // user: %6
  %6 = partial_apply [callee_guaranteed] %5<Ï„_0_0>(%4) : $@convention(thin) <Ï„_0_0 where Ï„_0_0 : LearningPhaseDependent> (@in_guaranteed Ï„_0_0.Output.TangentVector, @guaranteed @callee_guaranteed (@in_guaranteed Ï„_0_0.Output.TangentVector) -> (@out Ï„_0_0.TangentVector, @out Ï„_0_0.Input.TangentVector)) -> (@out Ï„_0_0.Input.TangentVector, @out Ï„_0_0.TangentVector) // user: %7
  return %6 : $@callee_guaranteed (@in_guaranteed Ï„_0_0.Output.TangentVector) -> (@out Ï„_0_0.Input.TangentVector, @out Ï„_0_0.TangentVector) // id: %7
} // end sil function 'AD__$s27FastaiNotebook_07_batchnorm22LearningPhaseDependentPAAE7forwardy6OutputQz5InputQzF__vjp_src_0_wrt_0_1'

swift: /swift-base/swift/lib/SIL/SILFunction.cpp:74: static swift::SILFunction *swift::SILFunction::create(swift::SILModule &, swift::SILLinkage, llvm::StringRef, swift::CanSILFunctionType, swift::GenericEnvironment *, Optional<swift::SILLocation>, swift::IsBare_t, swift::IsTransparent_t, swift::IsSerialized_t, swift::ProfileCounter, swift::IsDynamicallyReplaceable_t, swift::IsExactSelfClass_t, swift::IsThunk_t, swift::SubclassScope, swift::Inline_t, swift::EffectsKind, swift::SILFunction *, const swift::SILDebugScope *): Assertion `!entry->getValue() && "function already exists"' failed.
Stack dump:
0.	Program arguments: /swift-tensorflow-toolchain/usr/bin/swift -frontend -c /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/00_load_data.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/01_matmul.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/01a_fastai_layers.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/02_fully_connected.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/02a_why_sqrt5.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/03_minibatch_training.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/04_callbacks.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/05_anneal.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/05b_early_stopping.swift /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/06_cuda.swift -primary-file /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/07_batchnorm.swift -emit-module-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/FastaiNotebook_07_batchnorm.build/07_batchnorm~partial.swiftmodule -emit-module-doc-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/FastaiNotebook_07_batchnorm.build/07_batchnorm~partial.swiftdoc -emit-module-source-info-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/FastaiNotebook_07_batchnorm.build/07_batchnorm~partial.swiftsourceinfo -emit-dependencies-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/FastaiNotebook_07_batchnorm.build/07_batchnorm.d -emit-reference-dependencies-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/FastaiNotebook_07_batchnorm.build/07_batchnorm.swiftdeps -target x86_64-unknown-linux-gnu -disable-objc-interop -I /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug -enable-testing -g -module-cache-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/ModuleCache -swift-version 4.2 -Onone -D SWIFT_PACKAGE -D DEBUG -enable-anonymous-context-mangled-names -parse-as-library -module-name FastaiNotebook_07_batchnorm -o /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/FastaiNotebook_07_batchnorm.build/07_batchnorm.swift.o -index-store-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux-gnu/debug/index/store -index-system-modules 
1.	Swift version 5.1.1-dev (Swift 571e9d9692)
2.	While running pass #4630 SILModuleTransform "Differentiation".
3.	While processing // differentiability witness for LearningPhaseDependent.forward(_:)
sil_differentiability_witness [serialized] [parameters 0 1] [results 0] @$s27FastaiNotebook_07_batchnorm22LearningPhaseDependentPAAE7forwardy6OutputQz5InputQzF : $@convention(method) <Self where Self : LearningPhaseDependent> (@in_guaranteed Self.Input, @in_guaranteed Self) -> @out Self.Output {
}

 on SIL function "@$s27FastaiNotebook_07_batchnorm22LearningPhaseDependentPAAE7forwardy6OutputQz5InputQzF".
 for 'forward(_:)' (at /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/07_batchnorm.swift:32:12)
4.	While creating SIL function "@AD__$s27FastaiNotebook_07_batchnorm22LearningPhaseDependentPAAE7forwardy6OutputQz5InputQzF__vjp_src_0_wrt_0_1".
 for 'gradForward(_:)' (at /fastai_dev/swift/FastaiNotebook_07_batchnorm/Sources/FastaiNotebook_07_batchnorm/07_batchnorm.swift:40:5)
```

The issue seems similar to https://github.com/apple/swift/pull/28621#issuecomment-562763390 and should be fixable with a patch to fastai/fastai_dev. I'll investigate.

rxwei commented:
Your test cases didn’t cover the scenario where duplicate ‘@derivative(of:)’s for the same function are defined in different files in the **same** Swift module.

rxwei commented:
Once the restriction is lifted, the following should compile without issues:
```swift
// File A.swift
func foo(x: Float) -> Float
// File B.swift
@derivative(of: foo)
private func foo(x: Float) -> (value: Float, differential: ...)
// File C.swift
@derivative(of: foo)
private func foo(x: Float) -> (value: Float, differential: ...)
```

Only derivative registrations in the same Swift module that have **conflicting access levels** should be diagnosed.

dan-zheng commented:
> Only derivative registrations in the same Swift module that have **overlapping access levels** should be diagnosed.

Thanks for sharing the nuance, that makes sense! Let's definitely add a test for this in the proper patch for TF-1021.

dan-zheng commented:
Confirmed that extended test suite passes. Merging!
